src/aleph/http/core.clj

```diff
@@ -480,112 +514,148 @@
 
-(defn send-chunked-file [ch ^HttpMessage msg ^HttpFile file]
+(defn send-chunked-file [^AtomicBoolean message-sent? ch ^HttpMessage msg ^HttpFile file]
   (let [raf (RandomAccessFile. ^File (.-fd file) "r")
         cf (ChunkedFile. raf
                          (.-offset file)
                          (.-length file)
                          (.-chunk-size file))]
     (try-set-content-length! msg (.-length file))
+    (.set message-sent? true)
     (netty/write ch msg)
     (netty/write-and-flush ch (HttpChunkedInput. cf))))
 
-(defn send-chunked-body [ch ^HttpMessage msg ^ChunkedInput body]
+(defn send-chunked-body [^AtomicBoolean message-sent? ch ^HttpMessage msg ^ChunkedInput body]
+  (.set message-sent? true)
   (netty/write ch msg)
   (netty/write-and-flush ch body))
 
-(defn send-file-region [ch ^HttpMessage msg ^HttpFile file]
+(defn send-file-region [^AtomicBoolean message-sent? ch ^HttpMessage msg ^HttpFile file]
   (let [raf (RandomAccessFile. ^File (.-fd file) "r")
         fc (.getChannel raf)
         fr (DefaultFileRegion. fc (.-offset file) (.-length file))]
     (try-set-content-length! msg (.-length file))
+    (.set message-sent? true)
     (netty/write ch msg)
     (netty/write ch fr)
     (netty/write-and-flush ch empty-last-content)))
 
-(defn send-file-body [ch ssl? ^HttpMessage msg ^HttpFile file]
+(defn send-file-body [^AtomicBoolean message-sent? ch ssl? ^HttpMessage msg ^HttpFile file]
   (cond
     ssl?
     (let [source (-> file
                      (bs/to-byte-buffers {:chunk-size 1e6})
                      s/->source)]
-      (send-streaming-body ch msg source))
+      (send-streaming-body message-sent? ch msg source))
 
     (chunked-writer-enabled? ch)
-    (send-chunked-file ch msg file)
+    (send-chunked-file message-sent? ch msg file)
 
     :else
-    (send-file-region ch msg file)))
+    (send-file-region message-sent? ch msg file)))
 
 (let [ary-class (class (byte-array 0))
 
+      handle-failure
+      (fn [f]
+        (when (instance? ChannelFuture f)
+          (.addListener ^ChannelFuture f ChannelFutureListener/CLOSE_ON_FAILURE)))
+
       ;; extracted to make `send-message` more inlineable
       handle-cleanup
       (fn [ch f]
         (-> f
-          (d/chain'
-            (fn [^ChannelFuture f]
-              (if f
-                (.addListener f ChannelFutureListener/CLOSE)
-                (netty/close ch))))
-          (d/catch' (fn [_]))))]
+            (d/chain'
+             (fn [^ChannelFuture f]
+               (if f
+                 (.addListener f ChannelFutureListener/CLOSE)
+                 (netty/close ch))))
+            (d/catch' (fn [_]))))
+
+      handle-exception
+      (fn [error-logger body e keep-alive? ^Channel ch]
+        (if (nil? error-logger)
+          (let [class-name (.getName (class body))]
+            (log/errorf "error sending body of type %s: %s"
+                        class-name
+                        (.getMessage ^Throwable e)))
```

Comment:
Do you ignore the stacktrace for this exception on purpose?

Reply:
ok, the exception is rethrowed later in the handler, so it makes sense in this particular case.